File: Go/fsrs_test.go
package fsrs

import (
	"math"
	"math/rand"
	"reflect"
	"testing"
	"time"
)

var testRand = rand.New(rand.NewSource(1))

func createDefaultScheduler() *Scheduler {
	scheduler, _ := NewScheduler(DefaultSchedulerConfig(), testRand)
	return scheduler
}

func TestNextInterval(t *testing.T) {
	var desiredRetentions []float64
	for i := 2; i <= 10; i++ {
		desiredRetentions = append(desiredRetentions, float64(i)/10.0)
	}

	expected := []int{34793, 2508, 387, 90, 27, 9, 3, 1, 1}

	config := DefaultSchedulerConfig()
	config.LearningSteps = []time.Duration{}
	config.EnableFuzzing = false
	config.MaximumInterval = math.MaxInt

	var actual []int
	for _, r := range desiredRetentions {
		config.DesiredRetention = r
		scheduler, _ := NewScheduler(config, testRand)
		interval := scheduler.CalculateNextReviewInterval(1.0)
		actual = append(actual, int(interval/dayDuration))
	}

	if !reflect.DeepEqual(expected, actual) {
		t.Errorf("Expected %v, but got %v", expected, actual)
	}
}

func TestFsrs(t *testing.T) {
	config := DefaultSchedulerConfig()
	config.LearningSteps = []time.Duration{}
	config.RelearningSteps = []time.Duration{}
	config.EnableFuzzing = false
	scheduler, _ := NewScheduler(config, testRand)
	card := NewCard(1)
	ratings := []Rating{Again, Good, Good, Good, Good, Good}
	var actualIntervals []int

	for _, rating := range ratings {
		reviewInterval := card.Interval
		card = scheduler.ReviewCard(card, rating, reviewInterval)
		actualIntervals = append(actualIntervals, int(card.Interval/dayDuration))
	}

	expectedIntervals := []int{1, 2, 6, 17, 44, 102}
	if !reflect.DeepEqual(expectedIntervals, actualIntervals) {
		t.Errorf("Expected intervals %v, but got %v", expectedIntervals, actualIntervals)
	}
}

func TestMemoState(t *testing.T) {
	parameters := []float64{0.6845422, 1.6790825, 4.7349424, 10.042885, 7.4410233, 0.64219797, 1.071918, 0.0025195254, 1.432437,
		0.1544, 0.8692766, 2.0696752, 0.0953, 0.2975, 2.4691248, 0.19542035, 3.201072, 0.18046261, 0.121442534}
	config := DefaultSchedulerConfig()
	config.Parameters = parameters
	scheduler, _ := NewScheduler(config, testRand)
	reviews := []struct {
		rating   Rating
		interval int
	}{
		{Again, 0},
		{Good, 1},
		{Good, 3},
		{Good, 8},
		{Good, 21},
	}

	finalCard1 := runReviews(scheduler, reviews)
	checkStabilityAndDifficulty(t, 31.722992, 7.382128, finalCard1)

	cardMod := Card{
		CardID:     1,
		Interval:   21 * dayDuration,
		Stability:  20.925528,
		Difficulty: 7.005062,
		State:      Review,
		Step:       0,
	}
	finalCard2 := scheduler.ReviewCard(cardMod, Good, cardMod.Interval)
	checkStabilityAndDifficulty(t, 40.87456, 6.9913807, finalCard2)
}

func TestMemoryState(t *testing.T) {
	scheduler := createDefaultScheduler()
	reviews := []struct {
		rating   Rating
		interval int
	}{
		{Again, 0},
		{Good, 0},
		{Good, 1},
		{Good, 3},
		{Good, 8},
		{Good, 21},
	}

	finalCard1 := runReviews(scheduler, reviews)
	checkStabilityAndDifficulty(t, 53.62691, 6.3574867, finalCard1)

	config2 := DefaultSchedulerConfig()
	parameters2 := make([]float64, len(config2.Parameters))
	copy(parameters2, config2.Parameters)
	parameters2[17] = 0.0
	parameters2[18] = 0.0
	parameters2[19] = 0.0
	config2.Parameters = parameters2
	scheduler2, _ := NewScheduler(config2, testRand)

	finalCard2 := runReviews(scheduler2, reviews)
	checkStabilityAndDifficulty(t, 53.335106, 6.3574867, finalCard2)
}

func TestGoodLearningSteps(t *testing.T) {
	scheduler := createDefaultScheduler()
	card := NewCard(1)
	if card.State != New {
		t.Errorf("Expected state New, but got %v", card.State)
	}

	card = scheduler.ReviewCard(card, Good, card.Interval)
	if card.State != Learning {
		t.Errorf("Expected state Learning, but got %v", card.State)
	}
	if card.Step != 1 {
		t.Errorf("Expected step 1, but got %v", card.Step)
	}
	if math.Abs(card.Interval.Minutes()-10.0) > 1e-9 {
		t.Errorf("Expected interval around 10 minutes, but got %v", card.Interval)
	}

	card = scheduler.ReviewCard(card, Good, card.Interval)
	if card.State != Review {
		t.Errorf("Expected state Review, but got %v", card.State)
	}
	if card.Interval < dayDuration {
		t.Errorf("Expected interval >= 1 day, but got %v", card.Interval)
	}
}

func TestAgainLearningSteps(t *testing.T) {
	scheduler := createDefaultScheduler()
	card := NewCard(1)
	card = scheduler.ReviewCard(card, Again, card.Interval)

	if card.State != Learning {
		t.Errorf("Expected state Learning, but got %v", card.State)
	}
	if card.Step != 0 {
		t.Errorf("Expected step 0, but got %v", card.Step)
	}
	if math.Abs(card.Interval.Minutes()-1.0) > 1e-9 {
		t.Errorf("Expected interval around 1 minute, but got %v", card.Interval)
	}
}

func TestLearningCardRateHardOneLearningStep(t *testing.T) {
	config := DefaultSchedulerConfig()
	config.LearningSteps = []time.Duration{10 * time.Minute}
	scheduler, _ := NewScheduler(config, testRand)
	card := NewCard(1)
	card = scheduler.ReviewCard(card, Hard, card.Interval)

	expectedInterval := time.Duration(10.0*1.5) * time.Minute
	if card.Interval != expectedInterval {
		t.Errorf("Expected interval %v, but got %v", expectedInterval, card.Interval)
	}
}

func TestNoLearningSteps(t *testing.T) {
	config := DefaultSchedulerConfig()
	config.LearningSteps = []time.Duration{}
	scheduler, _ := NewScheduler(config, testRand)
	card := NewCard(1)
	card = scheduler.ReviewCard(card, Again, card.Interval)

	if card.State != Review {
		t.Errorf("Expected state Review, but got %v", card.State)
	}
	if card.Interval < dayDuration {
		t.Errorf("Expected interval >= 1 day, but got %v", card.Interval)
	}
}

func TestMaximumInterval(t *testing.T) {
	config := DefaultSchedulerConfig()
	config.MaximumInterval = 100
	scheduler, _ := NewScheduler(config, testRand)
	card := NewCard(1)

	for range 10 {
		card = scheduler.ReviewCard(card, Easy, card.Interval)
	}

	if card.Interval > time.Duration(config.MaximumInterval)*dayDuration {
		t.Errorf("Interval %v exceeds maximum interval %v days", card.Interval, config.MaximumInterval)
	}
}

func TestStabilityLowerBound(t *testing.T) {
	scheduler := createDefaultScheduler()
	const stabilityMin = 0.001
	card := NewCard(1)

	for i := range 100 {
		nextReviewTime := card.Interval + dayDuration
		card = scheduler.ReviewCard(card, Again, nextReviewTime)
		if card.Stability < stabilityMin {
			t.Errorf("Stability %v is below lower bound %v on iteration %d", card.Stability, stabilityMin, i)
		}
	}
}

func runReviews(scheduler *Scheduler, reviews []struct {
	rating   Rating
	interval int
}) Card {
	card := NewCard(1)
	for _, review := range reviews {
		reviewInterval := time.Duration(review.interval) * dayDuration
		card = scheduler.ReviewCard(card, review.rating, reviewInterval)
	}
	return card
}

func checkStabilityAndDifficulty(t *testing.T, expectedStability, expectedDifficulty float64, card Card) {
	if math.Abs(expectedStability-card.Stability) > 1e-4 {
		t.Errorf("Expected stability %v, but got %v", expectedStability, card.Stability)
	}
	if math.Abs(expectedDifficulty-card.Difficulty) > 1e-4 {
		t.Errorf("Expected difficulty %v, but got %v", expectedDifficulty, card.Difficulty)
	}
}

File: Python/fsrs_test.py
import unittest
from dataclasses import replace
from datetime import timedelta

from fsrs import Card, Rating, Scheduler, State, DEFAULT_CONFIG, STABILITY_MIN


class FsrsTests(unittest.TestCase):

    def _check_stability_and_difficulty(self, card: Card, expected_stability: float, expected_difficulty: float):
        self.assertAlmostEqual(card.stability, expected_stability, places=4)
        self.assertAlmostEqual(card.difficulty, expected_difficulty, places=4)

    def test_next_interval(self):
        desired_retentions = [r / 10.0 for r in range(1, 11)]
        base_config = replace(DEFAULT_CONFIG, learning_steps=[], relearning_steps=[], maximum_interval=2**31 - 1, enable_fuzzing=False)
        actual_intervals = []
        for r in desired_retentions:
            current_config = replace(base_config, desired_retention=r)
            scheduler = Scheduler(current_config)
            interval = scheduler._calculate_next_review_interval(stability=1.0)
            actual_intervals.append(int(interval.days))
        expected_intervals = [3116769, 34793, 2508, 387, 90, 27, 9, 3, 1, 1]
        self.assertEqual(actual_intervals, expected_intervals)

    def test_fsrs(self):
        config = replace(DEFAULT_CONFIG, learning_steps=[], relearning_steps=[], enable_fuzzing=False)
        scheduler = Scheduler(config)
        card = Card(card_id=1)
        ratings = [Rating.AGAIN, Rating.GOOD, Rating.GOOD, Rating.GOOD, Rating.GOOD, Rating.GOOD]
        actual_intervals = []
        for rating in ratings:
            scheduler.review_card(card, rating, card.interval)
            actual_intervals.append(card.interval.days)
        self.assertEqual(actual_intervals, [1, 2, 6, 17, 44, 102])

    def test_memo_state(self):
        w = [0.6845422, 1.6790825, 4.7349424, 10.042885, 7.4410233, 0.64219797,
             1.071918, 0.0025195254, 1.432437, 0.1544, 0.8692766, 2.0696752,
             0.0953, 0.2975, 2.4691248, 0.19542035, 3.201072, 0.18046261, 0.121442534]
        config = replace(DEFAULT_CONFIG, w=w)
        scheduler = Scheduler(config)

        card = Card(card_id=1)
        reviews = [(Rating.AGAIN, 0), (Rating.GOOD, 1), (Rating.GOOD, 3), (Rating.GOOD, 8), (Rating.GOOD, 21)]
        for rating, elapsed_days in reviews:
            scheduler.review_card(card, rating, timedelta(days=elapsed_days))

        self._check_stability_and_difficulty(card, 31.722992, 7.382128)

        card.interval = timedelta(days=21)
        card.stability = 20.925528
        card.difficulty = 7.005062

        scheduler.review_card(card, Rating.GOOD, timedelta(days=21))
        self._check_stability_and_difficulty(card, 40.87456, 6.9913807)

    def test_memory_state(self):
        scheduler1 = Scheduler()
        card1 = Card(card_id=1)
        reviews = [(Rating.AGAIN, 0), (Rating.GOOD, 0), (Rating.GOOD, 1), (Rating.GOOD, 3), (Rating.GOOD, 8), (Rating.GOOD, 21)]
        for rating, elapsed_days in reviews:
            scheduler1.review_card(card1, rating, timedelta(days=elapsed_days))
        self._check_stability_and_difficulty(card1, 53.62691, 6.3574867)

        w2 = list(DEFAULT_CONFIG.w)
        w2[17:20] = [0.0, 0.0, 0.0]
        scheduler2 = Scheduler(replace(DEFAULT_CONFIG, w=w2))
        card2 = Card(card_id=1)
        for rating, elapsed_days in reviews:
            scheduler2.review_card(card2, rating, timedelta(days=elapsed_days))
        self._check_stability_and_difficulty(card2, 53.335106, 6.3574867)

    def test_good_learning_steps(self):
        scheduler = Scheduler()
        card = Card(card_id=1)

        scheduler.review_card(card, Rating.GOOD, card.interval)
        self.assertEqual(card.state, State.LEARNING)
        self.assertEqual(card.step, 1)
        self.assertAlmostEqual(card.interval.total_seconds() / 60, 10.0)

        scheduler.review_card(card, Rating.GOOD, card.interval)
        self.assertEqual(card.state, State.REVIEW)
        self.assertGreaterEqual(card.interval.days, 1.0)

    def test_again_learning_steps(self):
        scheduler = Scheduler()
        card = Card(card_id=1)
        scheduler.review_card(card, Rating.AGAIN, card.interval)
        self.assertEqual(card.state, State.LEARNING)
        self.assertEqual(card.step, 0)
        self.assertAlmostEqual(card.interval.total_seconds() / 60, 1.0)

    def test_learning_card_rate_hard_one_learning_step(self):
        config = replace(DEFAULT_CONFIG, learning_steps=[timedelta(minutes=10.0)])
        scheduler = Scheduler(config)
        card = Card(card_id=1)
        scheduler.review_card(card, Rating.HARD, card.interval)
        expected_interval = timedelta(minutes=10.0 * 1.5)
        self.assertAlmostEqual(card.interval.total_seconds(), expected_interval.total_seconds(), delta=1.0)

    def test_no_learning_steps(self):
        config = replace(DEFAULT_CONFIG, learning_steps=[])
        scheduler = Scheduler(config)
        card = Card(card_id=1)
        scheduler.review_card(card, Rating.AGAIN, card.interval)
        self.assertEqual(card.state, State.REVIEW)
        self.assertGreaterEqual(card.interval.days, 1.0)

    def test_maximum_interval(self):
        config = replace(DEFAULT_CONFIG, maximum_interval=100)
        scheduler = Scheduler(config)
        card = Card(card_id=1)
        for _ in range(10):
            scheduler.review_card(card, Rating.EASY, card.interval)
        self.assertLessEqual(card.interval.days, 100)

    def test_stability_lower_bound(self):
        scheduler = Scheduler()
        card = Card(card_id=1)
        for _ in range(100):
            next_review_time = card.interval + timedelta(days=1)
            scheduler.review_card(card, Rating.AGAIN, next_review_time)
            self.assertGreaterEqual(card.stability, STABILITY_MIN)


if __name__ == "__main__":
    unittest.main()
